### PR TITLE
Autocast support for cudnn RNNs

### LINKED ISSUE #36428: AMP autocast error in cnn-lstm forward
## 🐛 Bug

get cuDNN error: CUDNN_STATUS_BAD_PARAM in cnn-lstm network forward method

## To Reproduce

```
import torch
from torch import nn, optim
from torch.cuda.amp import GradScaler, autocast

class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.conv1 = nn.Conv1d(1, 10, 2)
        self.lstm = nn.LSTM(input_size=10, hidden_size=10,
                                batch_first=True,bidirectional=True)

    def forward(self, x):
        x = self.conv1(x)
        x = x.permute(0, 2, 1)
        x, _ = self.lstm(x)
        return x


net = Net().cuda()
net.train()
optimizer = optim.SGD(net.parameters(), lr=1e-3)

scaler = GradScaler()

optimizer.zero_grad()

with autocast():
    output = net(torch.rand(10,1,10).cuda())
```

## Environment
torch==1.6.0 nightly (2020-04-10)

@mcarilli 

cc @csarofeen @ptrblck

### PR DESCRIPTION
**Pull request by mcarilli**
Should close https://github.com/pytorch/pytorch/issues/36428.

The cudnn RNN API expects weights to occupy a flat buffer in memory with a particular layout.  This PR implements a "speed of light" fix:  [`_cudnn_rnn_cast_reflatten`](https://github.com/pytorch/pytorch/pull/42385/files#diff-9ef93b6a4fb5a06a37c562b83737ac6aR327) (the autocast wrapper assigned to `_cudnn_rnn`) copies weights to the right slices of a flat FP16 buffer with a single read/write per weight (as opposed to casting them to FP16 individually then reflattening the individual FP16 weights, which would require 2 read/writes per weight).

It isn't pretty but IMO it doesn't make rnn bindings much more tortuous than they already are.

The [test](https://github.com/pytorch/pytorch/pull/42385/files#diff-e68a7bc6ba14f212e5e7eb3727394b40R2683) tries a forward under autocast and a backward for the full cross product of RNN options and input/weight/hidden dtypes.  As for all FP16list autocast tests, forward output and backward grads are checked against a control where inputs (including RNN module weights in this case) are precasted to FP16 on the python side.

Not sure who to ask for review, tagging @ezyang and @ngimel because Ed wrote this file (almost 2 years ago) and Natalia did the most recent major [surgery](https://github.com/pytorch/pytorch/pull/12600).

Side quests discovered:
- Should we update [persistent RNN heuristics](https://github.com/pytorch/pytorch/blob/dbdd28207c5cf6c4a35ceb1de0811c4812e8882c/aten/src/ATen/native/cudnn/RNN.cpp#L584) to include compute capability 8.0?  Could be another PR but seems easy enough to include.
- Many (maybe all?!) the raw cudnn API calls in [RNN.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cudnn/RNN.cpp) are deprecated in cudnn 8.  I don't mind taking the AI to update them since my mental cache is full of rnn stuff, but that would be a substantial separate PR.

### CODE REVIEW

**Comment by ezyang** [comment]
oh god I probably should review this XD

**Comment by mcarilli** [comment]
> oh god I probably should review this XD

abandon hope all ye who enter here.  Given the complexity of the code and the small number of people who have safaried through it, I made the test _exhaustive_ to keep us on track.  The best thing I can say about my implementation is "if its stupid and it works it aint stupid."

**Review by ngimel — COMMENTED** [review]
Sorry, can't continue today.

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp:640]: It's still a big switch statement :( https://github.com/pytorch/pytorch/blob/ae67f4c8b89547d648b66a49048ce13982ed6be7/c10/core/ScalarType.h#L163-L179

If there's an alternative fast way to get the `at::ScalarType` from a `TensorOptions` i'd love to hear it.

Comment (ngimel) [reply aten/src/ATen/native/cudnn/RNN.cpp:640]: You are not constructing TypeMetas at runtime (like it used to be), a few comparisons in the switch statement should be fast.

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:363]: Haha it just so happens that the way the code calling `at::_cudnn_rnn` (which dispatches into my wrapper) works, the incoming `weight_buf_opt` always `has_value()`, and the value is either a defined or undefined Tensor.  See my [comment ](https://github.com/pytorch/pytorch/pull/42385/files#diff-85792c9e6600ba236d0db7382f29108fR1292-R1296) on `try_get_weight_buf` in `_cudnn_impl` right before the `at::_cudnn_rnn` call.

You're still right, I should be more careful here and not dereference unless it has_value().

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:357]: Yes, will add comment.

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp:640]: oops, linked `scalarTypeToTypeMeta` at first.  edited back to `typeMetaToScalarType`.  Line 184 looks ominously like constructing a bunch of things at runtime.

File: aten/src/ATen/autocast_mode.cpp
Comment (ezyang) [inline aten/src/ATen/autocast_mode.cpp:16]: Danger danger will robinson. `ATen/autocast_mode.cpp` is compiled as part of ATen_cpu and it should not access any headers in the CUDA directory. You will probably have to chuck these autocast wrappers in a separate file in `ATen/cuda`

File: aten/src/ATen/autocast_mode.cpp
Comment (ezyang) [reply aten/src/ATen/autocast_mode.cpp:491]: No, it shouldn't be necessary to use `hacky_wrapper_for_legacy_signatures`. cc @smessmer

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (ezyang) [inline aten/src/ATen/native/cudnn/RNN.cpp:623]: Any substantive change to logic here?

File: aten/src/ATen/autocast_mode.cpp
Comment (smessmer) [reply aten/src/ATen/autocast_mode.cpp:491]: Please don't use the hacky_wrapper. Maybe I have to make the name even more scary to prevent people from using it :P

Are you rebased on current master? You shouldn't need the hacky wrapper here if your PR base is recent enough.

Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:491]: recent-ish master. autogenned code in VariableTypeWhatever.cpp used `hacky_wrapper_for_legacy_signatures` for _cudnn_rnn.  will rebase and try without it.

Comment (ezyang) [reply aten/src/ATen/autocast_mode.cpp:491]: @smessmer see, this is why you should fix earlier rather than later. People still looking at codegen to figure out what to do ;)

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (ezyang) [inline aten/src/ATen/native/cudnn/RNN.cpp:468]: This condition seems a bit backwards. Wouldn't `include_bias || cudnn_method != cudnnGetRNNLinLayerBiasParams` be more direct?

**Comment by ezyang** [comment]
While I am still not sure why you had to factor out a chunk of code into a separate helper, overall the changes seem reasonable and lightweight. Happy to approve when this is out of WIP.

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp:468]: works for me.

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp:640]: What's the verdict on this?

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp:623]: Added knobs are to make sure it can service both [_cudnn_rnn_flatten_weight](https://github.com/pytorch/pytorch/pull/42385/files#diff-85792c9e6600ba236d0db7382f29108fR698) and [autocast::_cudnn_rnn_cast_reflatten](https://github.com/pytorch/pytorch/pull/42385/files#diff-9ef93b6a4fb5a06a37c562b83737ac6aR336).

The existing behavior of `_cudnn_rnn_flatten_weight` should remain unchanged.

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:16]: hmmm but if I move my wrapper as an inline function in a header (or header + cpp) in ATen/cuda, then include the header in autocast_mode.cpp, seems like I'm no better off.

Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:16]: Do you mean I should put my wrapper declaration+definition in, say, `ATen/cuda/AutocastRNN.h`+`.cpp`,
have `ATen/autocast_mode.cpp` include `AutocastRNN.h`,
and have `ATen/cuda/AutocastRNN.cpp` be the thing that includes `ATen/native/cudnn/RNNUtils.h`?
That would mean `ATen/native/cudnn/RNNUtils.h` doesn't wind up directly included in `autocast_mode.cpp`.  Or am I misinterpreting?

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:361]: hopefully resolved

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp]: hopefully resolved

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp]: hopefully resolved

File: aten/src/ATen/native/RNN.h
Comment (mcarilli) [reply aten/src/ATen/native/RNN.h]: hopefully resolved

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp]: hopefully resolved

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp]: hopefully resolved

File: aten/src/ATen/autocast_mode.cpp
Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp:16]: Also, are the files in `ATen/cuda` compiled into a separate library from ATen_cpu?  If so, and I do the above, should I declare `_cudnn_rnn_cast_reflatten` with `TORCH_CUDA_API`?

Would doing the above using `ATen/cudnn` instead of `ATen/cuda` to host my files be equally valid?  If so,  I think `cudnn` rather than `cuda` is a more appropriate home.

Comment (ezyang) [reply aten/src/ATen/autocast_mode.cpp:16]: You should do the registration inside the cpp file in cuda. Then you do not need to include the header from `ATen/autocast_mode.cpp`

Comment (ezyang) [reply aten/src/ATen/autocast_mode.cpp:16]: Feel free to put it in cudnn directory; both dirs end up in torch_cuda library in the end.

Comment (mcarilli) [reply aten/src/ATen/autocast_mode.cpp]: hopefully resolved

File: test/test_cuda.py
Comment (ezyang) [inline test/test_cuda.py:2611]: cc @mruberry for an interesting ad hoc testing example

**Review by ezyang — APPROVED** [review]

**Comment by ezyang** [comment]
```

19:36:09 ======================================================================
19:36:09 FAIL: test_autocast_rnn (__main__.TestCuda)
19:36:09 ----------------------------------------------------------------------
19:36:09 Traceback (most recent call last):
19:36:09   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 815, in wrapper
19:36:09     method(*args, **kwargs)
19:36:09   File "test_cuda.py", line 2659, in test_autocast_rnn
19:36:09     self.assertEqual(out.grad_fn.name(), "CudnnRnnBackward")
19:36:09   File "/var/lib/jenkins/.local/lib/python3.6/site-packages/torch/testing/_internal/common_utils.py", line 1126, in assertEqual
19:36:09     super().assertEqual(x, y, msg=msg)
19:36:09 AssertionError: 'CatBackward' != 'CudnnRnnBackward'
19:36:09 - CatBackward
19:36:09 + CudnnRnnBackward
```

rocm failure is real

also lint error

```
  {
    path: 'test/test_cuda.py',
    start_line: 2610,
    end_line: 2610,
    start_column: 5,
    end_column: 5,
    annotation_level: 'failure',
    message: '[E301] expected 1 blank line, found 0'
  }
```

**Comment by mcarilli** [comment]
Thanks for quick review!  Users will be happy about this.

oops, I left the lint failure deliberately as a reminder to discuss if [the test should probably be tried without cudnn as well](https://github.com/pytorch/pytorch/pull/42385/files#diff-e68a7bc6ba14f212e5e7eb3727394b40R2603-R2604).  What do you think?

As for rocm, are we ok to `@skipIfRocm`, or should the test be tried with rocm?
Tbh I'm confused why the existing ` @unittest.skipIf(not TEST_CUDNN, 'CUDNN not available')` allowed the to run with rocm, is TEST_CUDNN considered true for rocm??

**Comment by ngimel** [comment]
Yes, `TEST_CUDNN = TEST_CUDA and (TEST_WITH_ROCM or something else)`.  And yes, it's ok to @skipIfRocm

**Comment by mcarilli** [comment]
Failures look spurious now (rocm failure is in test_nn.py)

File: aten/src/ATen/native/cudnn/RNN.cpp
Comment (mcarilli) [reply aten/src/ATen/native/cudnn/RNN.cpp:640]: On Slack, Natalia says 
> Just checked TypeMetaToScalarType for my PR, it’s...better avoided
so I'll keep working around it.

Comment (ngimel) [reply aten/src/ATen/native/cudnn/RNN.cpp:640]: I mean, compared to 100 `.size()` calls that RNN is making regardless it's nothing, but yeah, from the purist perspective you might as well not call it here.

**Comment by ezyang** [comment]
OK, I need to lock changes to this PR from OSS side, as it looks like this PR needs fbcode side build system changes.

**Pull request closed without merging**
